### PR TITLE
Migrate to hatch

### LINKED ISSUE #1510: [ENH] - Migrate to Hatch
### Feature description

Since we already have a `pyproject.toml` it would be good to migrate to [hatch](https://hatch.pypa.io/latest/) for dev and build bits. 

### Value and/or benefit

Hatch is pretty intuitive and allows for custom plugins to be added so this might be beneficial for some other enhancements planned in Nebari

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by iameskild**
Fixes #1510 

## Changes introduced in this PR:

- Migrate build-system to hatchling
- Move all project configs to `pyproject.toml`
  - flake8 is the one exception unfortunately - [see relevant issue](https://github.com/PyCQA/flake8/issues/234)
- Move black, flake8, and codespell to `.pre-commit-config.yaml`
  - Pre-commit also appears to also not respect `pyproject.toml` :(

## Types of changes

What types of changes does your PR introduce?

_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Build related changes
- [ ] Other (please describe):

### Requires testing

- [x] Yes
- [ ] No

### In case you checked yes, did you write tests?

- [ ] Yes
- [x] No

## Documentation

Does your contribution include breaking changes or deprecations?
If so have you updated the documentation?

- [ ] Yes, docstrings
- [ ] Yes, main documentation
- [ ] Yes, deprecation notices

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered and more.

### CODE REVIEW

**Review by costrouc — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -20,23 +20,36 @@ repos:
...
 # python
 - repo: https://github.com/psf/black
   rev: 22.3.0
   hooks:
   - id: black
+    args: [
+      "--line-length=88",
+      "--target-version=py38",
+      "--target-version=py39",
+      "--target-version=py310",
```
Comment (trallard) [inline .pre-commit-config.yaml:40]: this can be removed

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -32,7 +32,6 @@ jobs:
     strategy:
       matrix:
         python-version:
-          - "3.7"
           - "3.8"
           - "3.9"
           - "3.10"
```
Comment (trallard) [inline .github/workflows/test.yaml:37]: As agreed need to start testing on `3.11`

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -20,23 +20,36 @@ repos:
...
 
 - repo: https://github.com/pycqa/flake8
   rev: 3.8.4
   hooks:
   - id: flake8
     args: [
-      "--builtins=c"
+      "--builtins=c",
+      "--ignore=E203,E266,E501,W503",
+      "--max-line-length=89",
```
Comment (trallard) [inline .pre-commit-config.yaml:51]: let's remove this and leave all style to black

**Review by trallard — CHANGES_REQUESTED** [review]

**Comment by trallard** [comment]
Discussed with @iameskild  to add the creation of dev environments like we do in the hub theme

**Comment by ofek** [comment]
Hatch maintainer here! Nice 🙂 FYI flake8 can now be replaced by [Ruff](https://github.com/charliermarsh/ruff):

- https://github.com/bokeh/bokeh/pull/12605
- https://github.com/zulip/zulip/pull/23431
- https://github.com/pydantic/pydantic/pull/4709
- https://twitter.com/tiangolo/status/1591912354882764802
- https://github.com/pypa/hatch/pull/607 & https://github.com/pypa/hatch/pull/608
- ...

**Comment by trallard** [comment]
@iameskild  is there anything missing on your end or is this ready for a last review?

**Review by costrouc — APPROVED** [review]
@iameskild this looks good to me. I'm finally happy with the direction that version numbers are going instead of relying on version control.

**Comment by iameskild** [comment]
@trallard this is ready for a final review. I'm not sure where the [`pre-commit.ci` ](https://results.pre-commit.ci/run/github/252841745/1671040626.PBvte5NBSSGejz3vol5hIg)came from and why its failing (the regular pre-commit GHA seems to be working as expected).

**Review by trallard — APPROVED** [review]
LGTM - I added a comment but it is non-blocking at all

**Pull request merged**